core/cloudflow-operator/src/main/scala/cloudflow/operator/event/ConfigInputChangeEvent.scala

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cloudflow.operator
+package event
+
+import java.nio.charset.StandardCharsets
+import scala.collection.JavaConverters._
+import scala.concurrent._
+import scala.util.Try
+
+import akka.actor._
+import akka.NotUsed
+import akka.stream.scaladsl._
+import com.typesafe.config._
+import skuber._
+import skuber.api.client._
+import skuber.json.format._
+
+import cloudflow.operator.action._
+
+/**
+ * Indicates that the configuration of the application has been changed by the user.
+ */
+case class ConfigInputChangeEvent(appId: String, namespace: String, watchEvent: WatchEvent[Secret])
+
+object ConfigInputChangeEvent {
+  val SecretDataKey = "secret.conf"
+
+  /** log message for when a ConfigInputChangeEvent is identified as a configuration change event */
+  def detected[T <: ObjectResource](event: ConfigInputChangeEvent) =
+    s"User created or modified configuration for application ${event.appId}."
+
+  /**
+   * Transforms [[skuber.api.client.WatchEvent]]s into [[ConfigInputChangeEvent]]s.
+   */
+  def fromWatchEvent()(implicit system: ActorSystem): Flow[WatchEvent[Secret], ConfigInputChangeEvent, NotUsed] =
+    Flow[WatchEvent[Secret]]
+      .statefulMapConcat { () ⇒
+        var currentSecrets = Map[String, WatchEvent[Secret]]()
+        watchEvent ⇒ {
+          val secret       = watchEvent._object
+          val metadata     = secret.metadata
+          val secretName   = secret.metadata.name
+          val namespace    = secret.metadata.namespace
+          val absoluteName = s"$namespace.$secretName"
+          val data         = getData(secret)
+
+          watchEvent._type match {
+            case EventType.DELETED ⇒
+              currentSecrets = currentSecrets - absoluteName
```

Comment:
given that this starts from `empty`, this clause will potentially not delete anything (unless it gets an `ADDED` and `DELETED` in the same message).
To me, it looks like it's missing the current state somewhere.
ie., in like 53, something like:
```
var currentSecrets = getCurrentSecrets(namespace)
```

Reply:
Te currentSecrets is kept in statefulMapConcat across messages, this will only delete if earlier added or modified, which is what is intended.

Reply:
Also note the forall after the get in Add / Modify, so add is done if no item is found.